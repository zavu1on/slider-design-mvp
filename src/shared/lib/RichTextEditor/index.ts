export { RichTextEditor } from './RichTextEditor';
export { useRichTextStore } from './useRichTextStore';
