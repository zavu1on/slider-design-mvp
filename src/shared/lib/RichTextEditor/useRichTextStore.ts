import type { Editor } from '@tiptap/core';
import { create } from 'zustand';

type HeadingLevel = 1 | 2 | 3 | 4 | 5 | 6;

interface RichTextState {
  editor: Editor | null;
  setEditor: (editor: Editor | null) => void;

  toggleBold: () => void;
  toggleItalic: () => void;
  toggleStrike: () => void;
  toggleUnderline: () => void;
  toggleHeading: (level: HeadingLevel) => void;
}

export const useRichTextStore = create<RichTextState>((set, get) => ({
  editor: null,

  setEditor: (editor) => set((state) => ({ ...state, editor })),

  toggleBold: () => {
    const editor = get().editor;
    if (editor) editor.chain().focus().toggleBold().run();
  },
  toggleItalic: () => {
    const editor = get().editor;
    if (editor) editor.chain().focus().toggleItalic().run();
  },
  toggleStrike: () => {
    const editor = get().editor;
    if (editor) editor.chain().focus().toggleStrike().run();
  },
  toggleUnderline: () => {
    const editor = get().editor;
    if (editor) editor.chain().focus().toggleUnderline().run();
  },

  toggleHeading: (level) => {
    const editor = get().editor;
    if (editor) {
      editor.chain().focus().toggleHeading({ level }).run();
    }
  },
}));
