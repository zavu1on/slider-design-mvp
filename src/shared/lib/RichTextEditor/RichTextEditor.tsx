'use client';

import { type FC, useEffect } from 'react';
import { EditorContent, useEditor } from '@tiptap/react';
import StarterKit from '@tiptap/starter-kit';
import { useRichTextStore } from './useRichTextStore';

type RichTextEditorProps = {
  initialContent: string;
  className?: string;
  onBlur?: (html: string) => void;
};

export const RichTextEditor: FC<RichTextEditorProps> = ({
  initialContent,
  className,
  onBlur,
}) => {
  const setEditor = useRichTextStore((store) => store.setEditor);

  const editor = useEditor({
    extensions: [StarterKit],
    content: initialContent,
    immediatelyRender: false,
  });

  useEffect(() => {
    setEditor(editor);
    return () => setEditor(null);
  }, [editor, setEditor]);

  return (
    <EditorContent
      editor={editor}
      className={className}
      onBlur={() => {
        if (editor && onBlur) onBlur(editor.getHTML());
      }}
    />
  );
};
