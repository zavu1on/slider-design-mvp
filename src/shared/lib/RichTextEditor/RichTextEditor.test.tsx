import { cleanup, fireEvent, render, screen } from '@testing-library/react';
import { beforeEach, describe, expect, it, vi } from 'vitest';
import { RichTextEditor } from './RichTextEditor';
import { useRichTextStore } from './useRichTextStore';

describe('RichTextEditor component', () => {
  beforeEach(() => {
    cleanup();
    useRichTextStore.setState({ editor: null });
  });

  it('renders EditorContent with initial content', () => {
    render(<RichTextEditor initialContent="<p>Hello</p>" />);
    const editor = screen.getByTestId('editor-content');
    expect(editor).toBeInTheDocument();
  });

  it('stores editor in Zustand store on mount and clears on unmount', () => {
    const { unmount } = render(<RichTextEditor initialContent="<p>Test</p>" />);
    expect(useRichTextStore.getState().editor).not.toBeNull();

    unmount();
    expect(useRichTextStore.getState().editor).toBeNull();
  });

  it('calls onBlur with HTML content', () => {
    const handleBlur = vi.fn();
    render(
      <RichTextEditor initialContent="<p>Blur me</p>" onBlur={handleBlur} />
    );

    const editor = screen.getByTestId('editor-content');
    fireEvent.blur(editor);

    expect(handleBlur).toHaveBeenCalledTimes(1);
    expect(handleBlur.mock.calls[0][0]).toContain('<p>');
  });

  it('applies className', () => {
    render(<RichTextEditor initialContent="" className="custom-class" />);
    expect(screen.getByTestId('editor-content')).toHaveClass('custom-class');
  });

  it('updates editor content when typing', () => {
    render(<RichTextEditor initialContent="<p></p>" />);
    const editor = useRichTextStore.getState().editor;

    editor?.commands.setContent('<p>Typed text</p>');

    expect(editor?.getHTML()).toContain('Typed text');
  });

  it('snapshot editor', () => {
    const { container } = render(
      <RichTextEditor initialContent="<p>Hello World!</p>" />
    );
    expect(container).toMatchSnapshot();
  });
});
