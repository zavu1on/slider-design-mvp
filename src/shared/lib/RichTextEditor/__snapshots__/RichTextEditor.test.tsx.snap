// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RichTextEditor component > snapshot editor 1`] = `
<div>
  <div
    data-testid="editor-content"
  >
    <div
      class="tiptap ProseMirror"
      contenteditable="true"
      role="textbox"
      tabindex="0"
      translate="no"
    >
      <p>
        Hello World!
      </p>
    </div>
  </div>
</div>
`;
